src/platform/input/game-pads.js

```diff
@@ -68,62 +166,763 @@ const MAPS = {
+            this.touched = current > 0;
+        } else {
+            this.value = current.value;
+            this.pressed = current.pressed;
+            this.touched = current.touched ?? current.value > 0;
+            this._button = current;
+        }
+
+        if (previous) {
+            if (typeof previous === 'number') {
+                this._previouslyPressed = previous === 1;
+                this._previouslyTouched = previous > 0;
+            } else {
+                this._previouslyPressed = previous.pressed;
+                this._previouslyTouched = previous.touched ?? previous.value > 0;
+            }
+        } else {
+            this.updatePrevious();
+        }
+    }
+
+    /**
+     * Update the existing GamePadButton Instance.
+     *
+     * @param {GamepadButton} button - The original Gamepad API gamepad button.
+     * @ignore
+     */
+    update(button) {
+        this.updatePrevious();
+
+        this.value = button.value;
+        this.pressed = button.pressed;
+        this.touched = button.touched ?? button.value > 0;
+    }
+
+    /**
+     * Update the previous values for this button.
+     *
+     * @ignore
+     */
+    updatePrevious() {
+        this._previouslyPressed = this.pressed;
+        this._previouslyTouched = this.touched;
+    }
+
+    /**
+     * Whether the button was pressed.
+     *
+     * @type {boolean}
+     */
+    get wasPressed() {
+        return !this._previouslyPressed && this.pressed;
+    }
+
+    /**
+     * Whether the button was released since the last update.
+     *
+     * @type {boolean}
+     */
+    get wasReleased() {
+        return this._previouslyPressed && !this.pressed;
+    }
+
+    /**
+     * Whether the button was touched since the last update.
+     *
+     * @type {boolean}
+     */
+    get wasTouched() {
+        return !this._previouslyTouched && this.touched;
+    }
+}
+
+/**
+ * A GamePad stores information about a gamepad from the Gamepad API.
+ */
+class GamePad {
+    /**
+     * Create a new GamePad Instance.
+     *
+     * @param {Gamepad} gamepad - The original Gamepad API gamepad.
+     * @param {object} map - The buttons and axes map.
+     * @hideconstructor
+     */
+    constructor(gamepad, map) {
+        /**
+         * The identifier for the gamepad. Its structure depends on device.
+         *
+         * @type {string}
+         */
+        this.id = gamepad.id;
+
+        /**
+         * The index for this controller. A gamepad that is disconnected and reconnected will retain the same index.
+         *
+         * @type {number}
+         */
+        this.index = gamepad.index;
+
+        /**
+         * The buttons present on the GamePad. Some buttons may be null. Order is provided by API, use GamePad#buttons instead.
+         *
+         * @type {GamePadButton[]}
+         * @ignore
+         */
+        this._buttons = gamepad.buttons.map((b) => {
+            return b ? new GamePadButton(b) : null;
+        });
```

Comment:
```suggestion
        this._buttons = gamepad.buttons.map((b) => {
            return new GamePadButton(b);
        });
```
Can a native controller button really be `null`? And if so, then the type would be `(GamePadButton|null)[]`.

Reply:
Yes it can be as per XR Gamepads documentation. And no the type definition in parenthesis is not supported, I already tried in previous commit but just did again and still failing...

Reply:
Can you read it again? It says that it exactly cannot be `null`:

> Devices that lack one of the optional inputs listed in the tables above MUST preserve their place in the [buttons](https://www.w3.org/TR/gamepad/#dom-gamepad-buttons) or [axes](https://www.w3.org/TR/gamepad/#dom-gamepad-axes) array, reporting a [placeholder button](https://immersive-web.github.io/webxr-gamepads-module/#placeholder-button) or [placeholder axis](https://immersive-web.github.io/webxr-gamepads-module/#placeholder-axis), respectively. A placeholder button MUST report 0 for [value](https://www.w3.org/TR/gamepad/#dom-gamepadbutton-value), false for [pressed](https://www.w3.org/TR/gamepad/#dom-gamepadbutton-pressed), and false for [touched](https://www.w3.org/TR/gamepad/#dom-gamepadbutton-touched). A placeholder axis MUST report 0. [Placeholder buttons](https://immersive-web.github.io/webxr-gamepads-module/#placeholder-button) and [axes](https://immersive-web.github.io/webxr-gamepads-module/#placeholder-axis) MUST be omitted if they are the last element in the array or all following elements are also [placeholder buttons](https://immersive-web.github.io/webxr-gamepads-module/#placeholder-button) or [axes](https://immersive-web.github.io/webxr-gamepads-module/#placeholder-axis).

Reply:
Oh right. I was more relying on the table with "required no" and miss that statement.